spec/requests/contributions_spec.rb

```diff
@@ -13,67 +13,90 @@
       location_attributes: { zip: '12e45' },
   }}
 
-  before { sign_in create(:user) }
+  let(:user) { create(:user) }
 
-  describe 'GET /index' do
-    it 'renders a successful response' do
-      create(:listing)
-      get contributions_url
-      expect(response).to be_successful
-    end
+  ['when logged in', 'when logged out but p2p is enabled'].each do |scenario|
+    describe scenario do
+      before do
+        sign_in user unless scenario =~ /logged out/
+        allow_any_instance_of(SystemSetting).to receive(:peer_to_peer?).and_return(true) if scenario =~ /p2p is enabled/
+      end
```

Comment:
In general, i don't love repeating buckets of examples across multiple contexts.

There are of course circumstances when that's appropriate, eg for cross-cutting concerns. But in many cases, it can quickly become difficult to know which examples _really_ care about being tested in all contexts and which just happened to be lumped in out of convenience. This then can hinder future refactoring.

In this case, iirc the key thing to test between the three contexts is whether `/contributions` is accessible. I'd probably make that clear in a separate `describe 'access'` or similar block where that's the only thing being asserted. The examples below deal with details of the page which i don't think need to be re-run across contexts that are about authorization.

All that said, this is just my gut feeling -- i'm certainly fine with leaving it as is.

Reply:
I've updated the tests as you've suggested